### ISSUE TITLE
ini_config / ding-lib missing library dependency for OSX 10102

### ISSUE THREAD

**Issue by santokum** [meetecho/janus-gateway#159]
I was just trying to compile this in OSX and found dependency missing for ini_config. Tried on both brew and port but couldn't find - not sure if anyone hit this before 
No package 'ini_config' found

**Comment by ploxiln** [comment]
It's not generally packaged for mac osx, since practically no osx software uses it (and until recently, janus couldn't build on osx due to missing `clock_gettime()`). As far as I can tell, the canonical location for ding-libs is https://fedorahosted.org/sssd/ and you can find a source tarball for ding-libs-0.4.0 on that page. You'll have to build and install it yourself.

**Comment by santokum** [comment]
Cheers. for clock_gettime() i found this useful https://ininjas.com/pro/pages/view/284/clock-gettime-in-ios-and-macosx particularly as lib and link it. I guess more issues with iniconfig specifically it has build dependency with libtdb-dev, i may need to build this too.

**Comment by lminiero** [comment]
As Pierce anticipated, clock_gettime and others were already addressed in #125

**Comment by lminiero** [comment]
Closing this issue as I assume the suggestion by Pierce was the solution.

**Comment by Chicooz** [comment]
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking for style of include used by make... GNU
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking build system type... x86_64-apple-darwin14.3.0
checking host system type... x86_64-apple-darwin14.3.0
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
checking if the linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm
checking the name lister (/usr/bin/nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-apple-darwin14.3.0 file names to x86_64-apple-darwin14.3.0 format... func_convert_file_noop
checking how to convert x86_64-apple-darwin14.3.0 file names to toolchain format... func_convert_file_noop
checking for /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... no
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking for dsymutil... dsymutil
checking for nmedit... nmedit
checking for lipo... lipo
checking for otool... otool
checking for otool64... no
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking for -force_load linker flag... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... yes
checking for gcc option to produce PIC... -fno-common -DPIC
checking if gcc PIC flag -fno-common -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin14.3.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for pkg-config... /usr/local/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for JANUS... no
configure: error: Package requirements (
                    glib-2.0 >= glib_version
                    nice
                    libmicrohttpd
                    jansson
                    libssl >= ssl_version
                    libcrypto
                    sofia-sip-ua
                    ini_config
                  ) were not met:

No package 'ini_config' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables JANUS_CFLAGS
and JANUS_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.



I have installed Pierce's suggestion clock_gettime() , I don't know how to configure ./configure --prefix=/opt/janus still gives the above 

thanks in advance for your help

**Comment by lminiero** [comment]
As explained in CONTRINUTING.md, please don't paste long lines of console text plainly in the issues or comments: refer to a pastebin link or something alike instead.

The suggestion the configure script gives about setting the JANUS_CFLAGS variables is misleading, as we don't have a hook in the configure script to use anything else than pkg-config to get info on where the library is. Did you install the library manually somewhere that's not the "usual" place? In general, even manual installations provide pkg-config metadata that allows pkg-config to figure out where the library is, but I'm not familiar with OSX so I wouldn't know how it works there.

I guess that I should try and write a simple handmade replacement of what ini_config provides, that is the parsing of INI configuration files: in principle it shouldn't be that hard, and that would prevent issues like these with libraries that are not so commonly available. Unfortunately I'm often too lazy to reinvent the wheel :-)

**Comment by Chicooz** [comment]
sorry about that Lorenzo,  I've updated the comment to remove unnecessary lines, 

I installed using this command : 

         gcc -dynamiclib clock_gettime.c -install_name /usr/lib/librt.dylib -compatibility_version 1.0 -current_version 1.0 -o librt.dylib
 

As explained in the  https://ininjas.com/pro/pages/view/284/clock-gettime-in-ios-and-macosx  

I dont know if this covers for the ini_config issue, but that what you have suggested earlier in this thread.

looking forward to get this solved, i am here to confirm anything needed in mac, i can also provide access  to my machine if you need that

**Comment by fel** [comment]
Hi. I believe I have managed to get everything compiling on OS X Yosemite. I hope this helps - I'm still yet to confirm that all of the demos work, but so far they have seemed to.

`libintl` seems to be a prerequisite and and comes along with `gettext` from homebrew; so we can just symlink it:

```
ln -s /usr/local/Cellar/gettext/0.19.4//include/libintl.* /usr/local/include
```

`ini_config` comes along with `ding-libs`. So to install that:

```
git clone https://git.fedorahosted.org/git/ding-libs.git
cd ding-libs
export CFLAGS=-I/usr/include/malloc
./configure
```
You'll then need to remove any `--version-script` LD flags from the Makefile (I believe this is only needed for ELF anyway). Then:

```
make
make install
```

This will then allow you to compile janus-gateway by pointing at the previously installed `ini_config`.

Taking this a bit further, I believe I have everything (including post processing) up and running with the following:

```
brew install curl libmicrohttpd jansson libnice srtp sofia-sip glib opus libogg gengetopt libwebsockets libusrsctp rabbitmq-c`

./configure --enable-post-processing --prefix=/opt/janus CFLAGS="-I/usr/local/include -lini_config -lnice" LIBCURL_LIBS="-L/usr/local/opt/curl/lib"  LIBCURL_CFLAGS="-I/usr/local/opt/curl/include -lcurl"
make
sudo make install
```

Note that I also needed to install a newer version of libnice than that with comes from homebrew to get UDP. Download a release greater than 0.1.7 (e.g. 0.1.13) and do the `./configure && make && make install` dance.
